### PR TITLE
Make formatting tasks run asynchronously

### PR DESCRIPTION
**Pull request by rchl**
- There are now three separate tasks, one for code actions on save,
   one for formatting request, and one for WillSaveWait request.
 - Those tasks are registered onto LspSaveCommand and run sequentially on saving.
 - Tasks by default have the timeout set to 1s (matching former sync request) and code actions to 2s and controlled by a setting.

~~I still need to clean this up and figure out the timeout stuff (timeout separate for each task or for whole save action?).~~

### CODE REVIEW

**Comment by rchl** [comment]
When using `lsp_save` command the `on_pre_save` and thus potentially `will_save` notification does not run at the right time. Will have to fix it somehow so that it runs before save tasks and is skipped before native save (only when `lsp_save` was used).

**Comment by rchl** [comment]
It should be ready now.
The way I handle `on_pre_save` might be a bit hacky but not that different from existing, similar listener hacks.

**Comment by rwols** [comment]
So can we remove the SyncRequestStatus class now?

I do think we should keep the lock, in case we implicitly rely on different threads mutating the `self._response_handlers` dictionary.

**Comment by rchl** [comment]
> So can we remove the SyncRequestStatus class now?

Should I do it here?

**Comment by rwols** [comment]
> Should I do it here?

Please do! :+1:

File: plugin/core/protocol.py [hunk 1]
```diff
@@ -152,9 +152,6 @@ class ErrorCode:
     RequestCancelled = -32800
     ContentModified = -32801
 
-    # Defined by us
-    Timeout = -40000
```
Comment (rwols) [inline plugin/core/protocol.py:156]: Nice catch.

**Pull request merged**